### PR TITLE
ip tagging: level compressed trie implementation

### PR DESCRIPTION
**Pull request by ccaraman**
Implementation of the Level Compressed Trie as described by the paper IP address lookup using LC-tries by Stefan Nelson and Gunnar Karlsson. 

For this implementation, I referenced https://github.com/beevek/libkrb/blob/master/krb/lc_trie.hpp which is based off of the C implementation found  at http://www.csc.kth.se/~snilsson/software/router/C/

Note: This doesn't support nested prefixes. I've added a TODO to add support in the next PR. 

*Risk Level*: Medium  - This will be the core logic behind the Ip Tagging Filter. 

*Testing*: Unit Tests. A todo has been added for performance benchmark. 

*Docs Changes*:  Docs will be updated when the filter consumes this implementation.
*Release Notes*: I'll update the release notes when the filter consumes this implementation. 

Related to #1001 

Next Pr's will do the following: 
- Consume the LC-Trie in the Ip Tagging filter and update the IP Tagging filter to consume V2.
- Add Prefix support
- Performance benchmark testing.

### CODE REVIEW

**Comment by mattklein123** [comment]
@ccaraman do you mind just marking down the notes we just went over so they don't get duplicated by other reviewers? Awesome stuff!

File: source/common/network/lc_trie.h [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+          ip_prefixes_.push_back(tag_data[i]);
+        }
+      }
+
+      // The trie_ vector can have at most twice the number of nodes in the ip_prefixes vector.
+      // However, due to the fill factor a buffer is added. Preallocate the size of the vector to
+      // avoid resizing while building the trie.
+      // TODO(ccaraman): Define a better buffer value for the trie_ size.
+      // TODO(ccaraman): Define a better buffer value for the trie_ size.
+      // The value(2000000) was used in the original implementation in
```
Comment (ccaraman) [inline source/common/network/lc_trie.h]: @brian-pane Do you have any insight as to why `2000000` was chosen? Is there a better way to compute a buffer given the fill_factor?

**Comment by ccaraman** [comment]
@junr03 Please take a first pass.

File: source/common/network/lc_trie.h (hunk 1 above)
Comment (brian-pane) [reply source/common/network/lc_trie.h]: In the original implementation, I think the author just picked 2000000 as a really big number that they thought would be big enough to handle any situation.  And then, after populating the buffer, they resized it down to the amount of space they actually ended up using.

I don't have a good way to compute the buffer size in advance, since it depends on the exact shape of the tree. However, I think it's okay to _not_ pre-size the buffer. I think std::vector uses a power-of-two allocator internally, so the number of resize operations will be O(log(N)) for N prefixes.  So my proposal is to just use std::vector, see how fast it is, and then optimize if needed.

File: source/common/network/lc_trie.h [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+   * @param  ip supplies the IP address searching for a prefix in string format.
+   * @return tag from the prefix that encompasses the input. An empty string is returned
+   *         if no prefix in the LC Trie exists.
+   */
+  std::string search(Network::Address::InstanceConstSharedPtr ip_address) const;
+
+private:
+  /**
+   * Extract n bits from input starting at position p.
+   * @param position supplies the position.
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: It's called `position` in the comment but `p` in the declaration.

File: source/common/network/lc_trie.h [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+   * @param n supplies the number of bits to extract.
+   * @param input supplies the
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: ... "IP address from which to extract the bits," maybe?

File: source/common/network/lc_trie.h [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+   */
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)>
+  static IpType extractBits(int p, int n, IpType input) {
+    // The IP's are stored in network byte order. By shifting the value to the left by p bits, the
+    // bits between 0 and p-1 are zero'd out. Then to get the n bits, shift the IP back by the
+    // address_size minus the number of desired bits.
+    return input << p >> (address_size - n);
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: I think this may be unsafe. C doesn't specify whether a right shift is logical (high order bits get filled with zeros) or logical (high order bits get filled with a copy of whatever the input high order bit was).

It's probably safest to bitwise-and the result of these shifts with a mask consisting of _n_ ones in the low order bits.

File: source/common/network/lc_trie.h [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+   * @param input supplies the
+   */
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)>
+  static IpType removeBits(int n, IpType input) {
+    // The IP's are stored in network byte order. By shifting the value to the left by n bits and
+    // back, this zeros out all bits between 0 and n-1.
+    return input << n >> n;
+  }
+
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)> struct IpPrefixes {
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: Should this struct be named "IpPrefix," singular, since it holds just one prefix?

File: source/common/network/lc_trie.h [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+    /**
+     * Level Compressed Trie (LC-Trie) implementation for IP(IPv4 or IPv6) prefixes.
+     *
+     * The C++ implementation of the LcTrie found at
+     * https://github.com/beevek/libkrb/blob/master/krb/lc_trie.hpp was used as a reference.
+     * The implemenation above is based off of S. Nilson's code at
+     * http://www.csc.kth.se/~snilsson/software/router/C/
+     *
+     * Note: This trie handles Ipv4 and Ipv6 separately.
+     * Note: The trie can only support up 524288(2^19) nodes. Please see LcNode for further
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: Would you be open to making this unlimited? The global routing table currently contains 700,000 prefixes, so it won't fit within the 2^19 limit. I know Envoy itself is unlikely to ever have that many prefixes configured; but if this implementation can support more nodes, it could be useful to people working on other projects someday.

File: source/common/network/lc_trie.h [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+      }
+    }
+
+    /**
+     * LcNode is a uint32_t. A wrapper is provided to simplify getting the branch, skip and
+     * address values held within the data structure.
+     *
+     * The LcNode has three parts to it
+     * - Branch: the first 5 bits represent the branching factor. This supports a branching factor
+     * of 2^31.
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: The max branching factor is just 31 rather than 2^31, right?

File: source/common/network/lc_trie.h [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+     * - Skip: the next 7 bits represent the number of bits to skip when looking at an IP address.
+     * This value can be between 0 and 127, so IPv6 is supported as well.
+     * - Address: the remaining 30 bits represent the pointer. If branch value is not zero, this
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: If I'm counting right, it seems like there are only 20 remaining bits.  32 bits minus 5 for the branching factor, minus 7 for the skip count.

File: source/common/network/lc_trie.h [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+     * Takes a skip value and shifts it itno the 6-12(inclusively) bits of uint32_t.
+     * @param node
+     * @return
+     */
+    static LcNode setSkip(LcNode node) { return node << 20; }
+
+    /**
+     * @return Returns the value at bits 6-12.
+     */
+    static LcNode getSkip(LcNode node) { return node >> 20 & 127; }
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: In order to make code like this error-prone, the mechanism I usually use is:
```
struct LcNode {
  uint32_t branch:5;
  uint32_t skip:7;
  uint32_t address:20;
};
```
Then you can just refer to fields in the struct by name, and let the compiler deal with all the bit shifting.

I don't know if bitfield structs like that are encouraged or discouraged in Envoy, though.

File: source/common/network/lc_trie.h (hunk 6 above)
Comment (brian-pane) [reply source/common/network/lc_trie.h]: I should note, though, that making this unlimited would require using a `uint64_t` instead of `uint32_t` for `LcNode`, so the LC trie would take up more cache space.

File: source/common/network/lc_trie.h [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+     * Builds the Level Compresesed Trie, by first sorting the base vector, removing nested prefixes
+     * and duplicated and then recursively builds the trie by invoking buildRecursively().
+     */
+    void buildLcTrie(std::vector<IpPrefixes<IpType>>& tag_data) {
+      if (tag_data.empty()) {
+        return;
+      }
+
+      // The node used for the trie uses the last 20 bits of uint32_t. The trie cannot support
+      // an input greater than 524288(2^19) nodes.
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: I think it might be possible for a trie built from 2^19 prefixes to end up with more than 2^20 nodes, if the fill_factor is set to something less than 0.5. Hmmm, maybe the safe limit is 2^20 * fill_factor, but I haven't thought through it fully, so please don't take my word for it. :)

File: source/common/network/lc_trie.h [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+      if (tag_data.size() > (1 << 19)) {
+        throw EnvoyException(fmt::format(
+            "Size of the input vector({0}) for LC-Trie is larger than the max size supported(2^19)",
+            tag_data.size()));
+      }
+
+      std::sort(tag_data.begin(), tag_data.end());
+      ip_prefixes_.push_back(tag_data[0]);
+
+      // Remove duplicate entries and nested prefixes.
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: Does the original LC trie code work with nested prefixes? I know the corresponding paper didn't use any nested prefixes in its examples, but it isn't clear to me if the design can handle nested prefixes without modifications.

File: test/common/network/lc_trie_test.cc [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+  for (const auto& kv : test_case) {
+    EXPECT_EQ(kv.second, trie.search(Utility::parseInternetAddress(kv.first)));
+  }
+
+  // Check the Ipv4 path doesn't crash when the Ipv4 trie is empty.
+  EXPECT_EQ("", trie.search(Utility::parseInternetAddress("1.2.3.4")));
+}
+
+TEST(LcTrie, BothIpvVersions) {
```
Comment (brian-pane) [inline test/common/network/lc_trie_test.cc]: In practice, will IPv4 and IPv6 need separate trees? I'm thinking about the case where a user has configured an IPv4 prefix with length 24, and then they get a request from an IPv6 address whose first 24 bits happen to match the prefix. In that scenario, the user will probably want the IPv6 address to _not_ match the IPv4 prefix.

**Comment by brian-pane** [comment]
Thanks for implementing this!

In case it's useful for unit tests, I computed the IPv4 CIDR prefixes corresponding to the examples in the original [paper](https://www.nada.kth.se/~snilsson/publications/IP-address-lookup-using-LC-tries/text.pdf). These prefixes probably don't exercise 100% of the trie-building code paths, but they do result in a trie with a mix of different branching factors (figure 3 from the paper).
```
0.0.0.0/4
16.0.0.0/4
40.0.0.0/5
64.0.0.0/3
96.0.0.0/4
112.0.0.0/4
128.0.0.0/3
160.0.0.0/6
164.0.0.0/6
168.0.0.0/5
176.0.0.0/5
184.0.0.0/5
192.0.0.0/3
232.0.0.0/8
233.0.0.0/8
```

File: source/common/network/lc_trie.h (hunk 9 above)
Comment (mattklein123) [reply source/common/network/lc_trie.h]: this is fine, SGTM.

File: test/common/network/lc_trie_test.cc (hunk 12 above)
Comment (brian-pane) [reply test/common/network/lc_trie_test.cc]: Oops, never mind; I just realized the LCTrie maintains two separate LCTrieInternal objects, so it won't accidentally match v6 addresses against v4 prefixes.

File: source/common/network/lc_trie.h [hunk 13]
```diff
@@ -0,0 +1,487 @@
...
+        k = 0;
+        // k is finding the number of entries in the base vector that have the same bit pattern.
+        while (p + k < first + n &&
+               static_cast<uint32_t>(extractBits<IpType, address_size>(
+                   output.prefix_, output.branch_, ip_prefixes_[p + k].ip_)) == bit_pattern) {
+          ++k;
+        }
+
+        if (k == 0) {
+          if (p == first + n) {
```
Comment (brian-pane) [inline source/common/network/lc_trie.h]: Do the unit tests currently cover both this and its "else" case?  I ask because the handling of the "k==0" situation (which requires interpolating the values for new child nodes that have to be created when the number of children for a node is less than 2^branch_) seems like one of the trickiest parts of the LC Trie construction.

File: source/common/network/lc_trie.h (hunk 11 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: I'll update the comment here and for the class and method definition. Currently, the code will throw an exception if there are nested prefixes. I have a todo to make the exception show which prefixes are nested. 

The future implementation for nested prefixes won't modify the algorithm or the underlying LcNode. There are two options to handle it: 
1. Have a sorted list of the nested prefixes 
2. Use another lc-trie for the nested prefixes
The most general prefix will have a way of accessing one of those options once it is implemented.

File: test/common/network/lc_trie_test.cc (hunk 12 above)
Comment (ccaraman) [reply test/common/network/lc_trie_test.cc]: I'll improve the comments for the LcTrie to make this clearer. Thank you for the feedback.

File: source/common/network/lc_trie.h (hunk 8 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Yes, that is a typo. I'll fix that.

File: source/common/network/lc_trie.h (hunk 1 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: That sounds good! I prefer your suggestion to having a magic number. I'll change the comment to say how this diverged from the reference implementation.

File: source/common/network/lc_trie.h (hunk 11 above)
Comment (mattklein123) [reply source/common/network/lc_trie.h]: FWIW, I would do the implementation that was done in the original C implementation. IMO it's very easy to understand and will perform very well (linked list of prefixes, accessed from trie node). It will also perform well for our case where 0 prefixes will be by far the most common case and 1 prefix will be the next most common case by a huge margin.

File: source/common/network/lc_trie.h (hunk 3 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Thank you for catching that. I'll update.

File: source/common/network/lc_trie.h (hunk 2 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Fixed.

File: source/common/network/lc_trie.h (hunk 7 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Yes, that is another typo (and compression of the logic 2^5-1 = 31).

File: test/common/network/lc_trie_test.cc [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    output.push_back(ip_tags);
+    ip_tags.second.clear();
+  }
+  return output;
+};
+} // namespace
+
+// TODO(ccaraman): Add a performance benchmark test.
+
+TEST(LcTrie, Ipv4) {
```
Comment (ccaraman) [inline test/common/network/lc_trie_test.cc]: I'm going to restructure these tests. There is lots of duplicate in logic, take in a list of strings for cidr range, build trie, search for an input.

File: source/common/network/lc_trie.cc [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        ip_prefix.ip_ = convertToIpv6(cidr_range.ip()->ipv6()->address());
+        ipv6_prefixes.push_back(ip_prefix);
+      }
+    }
+  }
+  ipv4_trie_.reset(new LcTrieInternal<Ipv4>(ipv4_prefixes, fill_factor, root_branching_factor));
+  ipv6_trie_.reset(new LcTrieInternal<Ipv6>(ipv6_prefixes, fill_factor, root_branching_factor));
+}
+
+std::string LcTrie::search(Network::Address::InstanceConstSharedPtr ip_address) const {
```
[the remark this answers is not in the document]
Comment (ccaraman) [reply source/common/network/lc_trie.cc]: fixed.

File: source/common/network/lc_trie.h (hunk 9 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Thank you for the suggestion. This is cleaner. I've updated the code to use your proposed structure.

File: source/common/network/lc_trie.h (hunk 6 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: I'm am open to it. I don't think we need it now. I've added a comment to change the data structure for LcNode to be uint64_t if any consumer requires it.

File: source/common/network/lc_trie.h (hunk 5 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Yes, that is clearer. I've updated it to be IpPrefix.

File: source/common/network/lc_trie.h (hunk 4 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: I double checked the c++ standard and for unsigned left/right shifting is safe and produces the values we expect. In this version of the pr, they were int's. I've change the methods to be uint32_t.

File: source/common/network/lc_trie.h (hunk 13 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: I double checked. The sample input provided in the paper covers all of the if/else cases in buildRecursive.

File: source/common/network/lc_trie.h [hunk 16]
```diff
@@ -0,0 +1,540 @@
...
+     * Note: If more than 2^19 CIDR ranges are to be stored in trie_, uint64_t should be used
+     * instead.
+     */
+    struct LcNode {
+      uint32_t branch_ : 5;
+      uint32_t skip_ : 7;
+      uint32_t address_ : 20;
+    };
+
+    // Vector that contains the CIDR ranges and tags.
```
Comment (junr03) [inline source/common/network/lc_trie.h]: explain that this is kept because you need to make sure that you've got a match. I know this is based on the paper, but it might be worth it to spell that here.

File: source/common/network/lc_trie.h [hunk 17]
```diff
@@ -0,0 +1,540 @@
...
+      if (n == 1) {
+        trie_[position].address_ = first;
+        return;
+      }
+
+      ComputePair output = computeBranch(prefix, first, n);
+
+      uint32_t address = next_free_index;
+      trie_[position].branch_ = output.branch_;
+      trie_[position].skip_ = output.prefix_ - prefix;
```
Comment (ccaraman) [inline source/common/network/lc_trie.h]: Synced with Jose offline - i'll add  a comment that the skip value is the number of bits between the last prefix at prefix and the output.prefix . (Mention that it is the number of bits)

File: source/common/network/lc_trie.h [hunk 18]
```diff
@@ -0,0 +1,540 @@
...
+      // The value of next_free_index is the final size of the trie_.
+      trie_.resize(next_free_index);
+    }
+
+    // Thin wrapper around computeBranch output to facilitate code readability.
+    struct ComputePair {
+      ComputePair(int branch, int prefix) : branch_(branch), prefix_(prefix) {}
+
+      uint32_t branch_;
+      uint32_t prefix_;
```
Comment (ccaraman) [inline source/common/network/lc_trie.h]: I'll add note that this is total number of bits that have the same prefix for a range of data.

File: source/common/network/lc_trie.h [hunk 19]
```diff
@@ -0,0 +1,540 @@
...
+
+    /**
+     * Compute the branch and skip values for the trie starting at position 'first' through
+     * 'first+n-1' while disregarding the prefix.
+     * @param prefix supplies the common prefix in the ip_prefixes_ array.
+     * @param first supplies the index where computing the branch should begin with.
+     * @param n supplies the number of nodes to use while computing the branch.
+     * @return pair of integers for the branching factor and the skip.
+     */
+    ComputePair computeBranch(uint32_t prefix, uint32_t first, uint32_t n) const {
```
Comment (junr03) [inline source/common/network/lc_trie.h]: I would rename this method. computeBranchAndSkip?

File: source/common/network/lc_trie.h (hunk 1 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: I decided to not remove the resizing due to the accessing pattern used across the code base. Once there are performance benchmark tests, the resizing the vector can be something we look to improve if needed.

File: source/common/network/lc_trie.h (hunk 6 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: To summarize our offline conversation, the code will stay with uint32_t as the underlying LcNode structure. 

To avoid getting an out of range exception, I'll add a check at the beginning of every buildRecursive call to ensure the size is > the position being accessed in the Lc-Trie that way if the number of trie_ nodes required is greater than 2^20  the code will throw an exception. If this case is it, then the uint64_t conversation should be revisited.

Once there are larger data sets and perf tests, there can be an effort to see if using uint64_t is worth supporting larger set of prefixes. Another benefit is then there can be some experimentation in determining what the buffer can be.

File: source/common/network/lc_trie.h [hunk 20]
```diff
@@ -0,0 +1,567 @@
...
+
+          if (pattern_found) {
+            ++count;
+          }
+          ++pattern;
+        }
+        // Stop iterating once the size of the branch (with the fill factor ratio) is able to
+        // contain all of the prefixes within the current range of ip_prefixes_[first] to
+        // ip_prefixes_[first+n-1].
+      } while (count >= fill_factor_ * (1 << branch));
```
Comment (junr03) [inline source/common/network/lc_trie.h]: I am a little concerned about the default fill factor of 0. That would mean that any count would keep increasing the branch going up and collapsing the trie

File: source/common/network/lc_trie.h [hunk 21]
```diff
@@ -0,0 +1,567 @@
...
+
+      // Set the branching factor.
+      compute.branch_ = branch - 1;
```
Comment (junr03) [inline source/common/network/lc_trie.h]: comment on the -1

File: source/common/network/lc_trie.h [hunk 22]
```diff
@@ -0,0 +1,567 @@
...
+                        uint32_t& next_free_index) {
+      // Setting a leaf, the branch and skip are 0.
+      if (n == 1) {
+        // There is no way to predictably determine the number of trie nodes required to build a
+        // LC-Trie. If while building the trie the position that is being set exceeds the maximum
+        // number of supported entries, throw an Envoy Exception instead of letting an out_of_range
+        // exception be thrown.
+        // TODO(ccaraman): Add a test case.
+        if (position > MAXIMUM_TRIE_NODES) {
+          throw EnvoyException(fmt::format("The number of internal nodes required for the LC-Trie "
```
Comment (junr03) [inline source/common/network/lc_trie.h]: I know this is not the responsibility of the throwing code. But given we own both the throwing code and the code that receives the exception, I want to make sure this is caught at config time, which I am 99% sure it does.

File: source/common/network/lc_trie.h [hunk 23]
```diff
@@ -0,0 +1,567 @@
...
+      // factor).
+      next_free_index += 1 << output.branch_;
+
+      uint32_t new_position = first;
+
+      // Build the subtrees.
+      for (uint32_t bit_pattern = 0; bit_pattern < static_cast<uint32_t>(1 << output.branch_);
+           ++bit_pattern) {
+
+        // count is the number of entries in the base vector that have the same bit pattern as the
```
Comment (junr03) [inline source/common/network/lc_trie.h]: s/base/ip_prefixes_

File: source/common/network/lc_trie.h [hunk 24]
```diff
@@ -0,0 +1,567 @@
...
+        int count = 0;
+        while (new_position + count < first + n &&
+               static_cast<uint32_t>(extractBits<IpType, address_size>(
+                   output.prefix_, output.branch_, ip_prefixes_[new_position + count].ip_)) ==
+                   bit_pattern) {
+          ++count;
+        }
+
+        // This logic was taken from
+        // https://github.com/beevek/libkrb/blob/24a224d3ea840e2e7d2926e17d8849aefecc1101/krb/lc_trie.hpp#L396.
```
Comment (junr03) [inline source/common/network/lc_trie.h]: I think this section lacks in explanation compared to the rest of the code, and it is a pretty crucial set of operations. I'll leave it up to you if you want to spell it out more. At the very least to explain what each case is.

File: source/common/network/lc_trie.h (hunk 20 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: fill factor will validated in the config to have this range (0, 1]

File: source/common/network/lc_trie.h (hunk 22 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Yes, this will get thrown during config parsing  The call chain is LcTrie()->LcTrieInternal<>-> build()->buildRecursive

File: source/common/network/lc_trie.h (hunk 23 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: fixed

File: source/common/network/lc_trie.h (hunk 24 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Added more comments

File: source/common/network/lc_trie.h [hunk 25]
```diff
@@ -0,0 +1,588 @@
...
+namespace Envoy {
+namespace Network {
+namespace LcTrie {
+
+/**
+ * Level Compressed Trie for tagging IP addresses. Both IPv4 and IPv6 addresses are supported
+ * within this class with no calling pattern changes.
+ *
+ * The algorithm to build the LC-Trie is desribed in the paper 'IP-address lookup using LC-tries'
+ * by'S. Nilsson' and 'G. Karlsson'. Refer to LcTrieInternal for implementation and algorithm
```
Comment (mattklein123) [inline source/common/network/lc_trie.h]: nit: missing space after "by". Please also provide direct link to paper/article if possible.

File: source/common/network/lc_trie.h [hunk 26]
```diff
@@ -0,0 +1,588 @@
...
+  /**
+   * @param tag_data supplies a vector of tag and CIDR ranges.
+   * @param fill_factor supplies the fraction of completeness to use when calculating the branch
+   *                    value for a sub-trie.
+   * @param root_branching_factor supplies the branching factor at the root. The paper suggests for
+   *                              large LC-Tries to use the value '16' for the root branching
+   *                              factor. It reduces the depth of the trie.
+   */
+  LcTrie(const std::vector<std::pair<std::string, std::vector<Address::CidrRange>>>& tag_data,
+         double fill_factor = 0.5, uint32_t root_branching_factor = 0);
```
Comment (mattklein123) [inline source/common/network/lc_trie.h]: Should we just make the default 16 for root_branching_factor?

File: source/common/network/lc_trie.h [hunk 27]
```diff
@@ -0,0 +1,588 @@
...
+  }
+
+  // IP addresses are stored in host byte order to simplify
+  typedef uint32_t Ipv4;
+  typedef absl::uint128 Ipv6;
+
+  // Helper methods to retrieve the string representation of the address in IpPrefix using
+  // inet_ntop. These strings are used in the nested prefixes exception messages.
+  // TODO(ccaraman): Remove once nested prefixes are supported.
+  static std::string toString(const Ipv4& input) {
```
Comment (mattklein123) [inline source/common/network/lc_trie.h]: For these can you just construct an Ipv4Instance or Ipv6Instance and return the string from that? Seems simpler.

File: source/common/network/lc_trie.h [hunk 28]
```diff
@@ -0,0 +1,588 @@
...
+              "LcTrie does not support nested prefixes. '{0}' is a nested prefix of '{1}'.",
+              tag_data[i].asString(), tag_data[i - 1].asString()));
+        }
+
+        if (tag_data[i - 1] != tag_data[i]) {
+          ip_prefixes_.push_back(tag_data[i]);
+        }
+      }
+
+      // In theory, the trie_ vector can have at most twice the number of  ip_prefixes entries - 1.
```
Comment (mattklein123) [inline source/common/network/lc_trie.h]: nit: extra space before ip_prefixes

File: source/common/network/lc_trie.h [hunk 29]
```diff
@@ -0,0 +1,588 @@
...
+    // Main trie search structure.
+    std::vector<LcNode> trie_;
+
+    const double fill_factor_;
+    const uint32_t root_branching_factor_;
+  };
+
+  /**
+   * Converts Ipv6 address in std::array<uint8_t, 16> to absl::uint128 in host byte order.
+   * TODO (ccaraman): Remove this workaround once Ipv6Helper returns absl::uint128t.
```
Comment (mattklein123) [inline source/common/network/lc_trie.h]: Can we just do that now? It doesn't seem that hard and seems like it would clean up a bunch of stuff. You could do another PR and just merge that into this?

Comment (ccaraman) [reply source/common/network/lc_trie.h]: I'll do this in a separate PR and merge it into here.

File: source/common/network/lc_trie.h (hunk 26 above)
Comment (brian-pane) [reply source/common/network/lc_trie.h]: I think it may be better to use a much smaller default, like 6 or 8.  Or maybe auto-compute the root branching factor based on the number of prefixes submitted to the build function.

Here's my rationale...

With a root_branching factor of 16, each trie will use at least 256KB of memory: 2^16 nodes at the first level below the root, times 4 bytes per node.

When Nilsson _et al_ recommended 16 bits of branching at the root in their [original paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.108.5093&rep=rep1&type=pdf), their target use case was a router with at least tens of thousands of prefixes. In that scenario, dedicating 256KB of memory to shard the trie into 2^16 buckets seems like a reasonable tradeoff of memory for speed.

But proxy applications, if my own experience is representative, will usually have a much smaller number of prefixes configured: often dozens of prefixes, sometimes hundreds, rarely tens of thousands. For such configurations, 256KB strikes me as too much overhead, especially relative to the L1D and even L2 cache sizes of contemporary server CPUs.

Comment (mattklein123) [reply source/common/network/lc_trie.h]: SGTM. Whatever you all think is best. Was mostly just wondering if the default should not be 0.

File: source/common/network/lc_trie.h (hunk 27 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Good idea. Fixed.

File: source/common/network/lc_trie.h (hunk 26 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: I'll leave it at zero for now.

File: source/common/network/lc_trie.h (hunk 25 above)
Comment (ccaraman) [reply source/common/network/lc_trie.h]: Fixed. Found a free version of the paper hosted by the author.

File: source/common/network/lc_trie.h [hunk 30]
```diff
@@ -0,0 +1,550 @@
...
+   * Retrieve the tag associated with the CIDR range that contains `ip_address`. Both IPv4 and IPv6
+   * addresses are supported.
+   * TODO(ccaraman): Change to getTags and std::vector<std::string> when nested prefixes are
+   * supported.
+   * @param  ip_address supplies the IP address.
+   * @return tag from the CIDR range that contains 'ip_address'. An empty string is returned
+   *             if no prefix contains 'ip_address' or there is no data for the IP version of the
+   *             ip_address.
+   */
+  std::string getTag(const Network::Address::InstanceConstSharedPtr ip_address) const;
```
Comment (mattklein123) [inline source/common/network/lc_trie.h]: nit: `const Network::Address::InstanceConstSharedPtr&`

File: test/common/network/lc_trie_test.cc [hunk 31]
```diff
@@ -0,0 +1,270 @@
...
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Network {
+namespace LcTrie {
+
+class LcTrieTest : public testing::Test {
+public:
+  LcTrieTest() {}
```
Comment (mattklein123) [inline test/common/network/lc_trie_test.cc]: nit: del
